geo/map_processing/map_structure.py

```diff
@@ -0,0 +1,212 @@
+        y, x = shapely_point.y, shapely_point.x
+        latlng = s2.S2LatLng.FromDegrees(y, x)
+        return s2.S2Polygon(s2.S2Cell(s2.S2CellId(latlng)))
+
+    def cellid_from_point(self, s2_point, level):
+        # OpenStreetMaps Nodes type to cellids
+        return s2.S2CellId(s2_point, level)
+
+    def s2point_from_coord_xy(self, coord):
+        latlng = s2.S2LatLng.FromDegrees(coord[1], coord[0])
+        return latlng.ToPoint()
+
+    def latlng_from_coord_xy(self, coord):
+        latlng = s2.S2LatLng.FromDegrees(coord[1], coord[0])
+        return latlng
+
+    def s2polygon_from_shapely_polygon(self, shapely_polygon):
+        if not hasattr(shapely_polygon.buffer(0.00005), 'exterior'):
+            return
+        else:
+            list_coords = list(shapely_polygon.buffer(0.00005).exterior.coords)
+
+        s2point_list = list(map(self.s2point_from_coord_xy, list_coords))
+        s2point_list = s2point_list[::-1]  # Counterclockwise
+        return s2.S2Polygon(s2.S2Loop(s2point_list))
+
+    def s2polygon_from_shapely_polyline(self, shapely_polygon):
+
+        list_coords = list(shapely_polygon.exterior.coords)
+
+        list_ll = []
+        for lat, lng in list_coords:
+            list_ll.append(s2.S2LatLng.FromDegrees(lat, lng))
+
+        line = s2.S2Polyline()
+        line.InitFromS2LatLngs(list_ll)
+
+        return line
+
+    def cellid_from_geometry(self, geo, level):
+        assert isinstance(geo, Point) or isinstance(geo, Polygon), type(geo)
+        if isinstance(geo, Point):
+            s2_polygon = self.s2polygon_from_shapely_point(geo)
+        else:
+            s2_polygon = self.s2polygon_from_shapely_polygon(geo)
+            if s2_polygon is None:
+                return None
+        return self.cellid_from_s2shape(s2_polygon, level)
+
+    def plot_cells(self, cells):
+        # create a map
+        map_osm = folium.Map(
+            location=[40.7434, -73.9847], zoom_start=12, tiles='Stamen Toner')
+
+        def style_function(x):
+            return {'weight': 1, 'fillColor': '#eea500'}
+
+        geoms = []
+        for cellid in cells:
+            cellid = cellid[0]
+            cell = s2.S2Cell(cellid)
+            vertices = []
+            for i in range(0, 4):
+                vertex = cell.GetVertex(i)
+
+                latlng = s2.S2LatLng(vertex)
+                vertices.append((latlng.lat().degrees(),
+                                 latlng.lng().degrees()))
+            gj = folium.GeoJson(
+                {
+                    "type": "Polygon",
+                    "coordinates": [vertices]
+                },
+                style_function=style_function)
+            gj.add_children(folium.Popup(cellid.ToToken()))
+            gj.add_to(map_osm)
+
+        filepath = 'visualization.html'
+        map_osm.save(filepath)
+        webbrowser.open(filepath, new=2)
+
+    def cellid_from_geometry_streets(self, geo, level):
```

Comment:
Why 'streets' here? Also, don't overload 'geo' here and create two functions instead (with types).

Reply:
changed